storage/imx6ull-flash/flashdev.c

```diff
@@ -451,6 +438,94 @@ static int flashmtd_metaWrite(struct _storage_t *strg, off_t offs, const void *d
 }
 
 
+/* MTD interface */
+
+static int flashmtd_erase(struct _storage_t *strg, off_t offs, size_t size)
+{
+	int res;
+	uint32_t pageID;
+	uint32_t eb, startBlock, endBlock, erased = 0; /* values in erase block units */
+	unsigned int pagesPerBlock;
+
+	if ((offs % strg->dev->mtd->erasesz) != 0 || (size % strg->dev->mtd->erasesz) != 0) {
+		return -EINVAL;
+	}
+
+	startBlock = offs / strg->dev->mtd->erasesz;
+	endBlock = (offs + size) / strg->dev->mtd->erasesz;
+	pagesPerBlock = strg->dev->mtd->erasesz / strg->dev->mtd->writesz;
+
+	mutexLock(strg->dev->ctx->lock);
+	for (eb = startBlock; eb < endBlock; ++eb) {
+		pageID = eb * pagesPerBlock;
+
+		res = flashdrv_isbad(strg->dev->ctx->dma, pageID);
+		if (res != 0) {
+			printf("flashmtd_erase: skipping bad block: %u\n", eb);
+			continue;
+		}
+
+		res = flashdrv_erase(strg->dev->ctx->dma, pageID);
+		if (res < 0) {
+			printf("flashmtd_erase: error while erasing block: %u - marking as badblock\n", eb);
+			res = flashdrv_markbad(strg->dev->ctx->dma, pageID);
+			if (res < 0) {
+				mutexUnlock(strg->dev->ctx->lock);
+				return res; /* can't mark as badblock, urecoverable error */
+			}
+		}
+		else {
+			++erased;
+		}
+	}
+	off_t start = offs - strg->start;
+	off_t end = start + size;
+	(void)cache_invalidate(strg->dev->ctx->dcache, start, end);
+	mutexUnlock(strg->dev->ctx->lock);
+
+	return erased;
+}
+
+
+static int flashmtd_read(struct _storage_t *strg, off_t offs, void *data, size_t len, size_t *retlen)
+{
+	mutexLock(strg->dev->ctx->lock);
+
+	off_t start = offs - strg->start;
+	int ret = cache_read(strg->dev->ctx->dcache, start, data, len);
+
+	mutexUnlock(strg->dev->ctx->lock);
+	if (ret >= 0) {
+		*retlen = len;
```

Comment:
not sure if this is is handled correctly, JFFS may get error (`EBADMSG`, maybe `EUCLEAN`) and still expects the data to be read (see https://github.com/phoenix-rtos/phoenix-rtos-devices/commit/8b1f8159b6ba247d371d682d0902169ceaaad4e8 for details).

Please investigate it in detail as AFAIK this was a fix for a long-running problem with incorrectly-overwritten data/metadata.

Reply:
Thanks for the comment, unfortunately it's not possible to return both error and number of read bytes through libcache interface so I'll be closing this.